src/dlfcn.c

```diff
@@ -467,7 +472,157 @@ char *dlerror( void )
     return error_buffer;
 }
 
-#ifdef SHARED
+/* taken from http://bandido.ch/programming/Import_Address_Table_Hooking.pdf */
+static IMAGE_IMPORT_DESCRIPTOR* getImportTable( HMODULE module, DWORD *size )
+{
+    IMAGE_DOS_HEADER* dosHeader = (IMAGE_DOS_HEADER*)module;
+    if ( dosHeader->e_magic != 0x5A4D )
+        return NULL;
+    IMAGE_OPTIONAL_HEADER* optionalHeader = (IMAGE_OPTIONAL_HEADER*)
+        ((BYTE*)module + dosHeader->e_lfanew + 24);
+    if ( optionalHeader->Magic != IMAGE_NT_OPTIONAL_HDR_MAGIC )
+        return NULL;
+    if ( optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size == 0 ||
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress == 0 )
+        return NULL;
+    *size = optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size;
+    return (IMAGE_IMPORT_DESCRIPTOR*)((BYTE*)module +
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress);
+}
+
+/*
+ * return symbol name for a given address
+ */
+static char *getSymbolName( HMODULE baseAddress, IMAGE_IMPORT_DESCRIPTOR *iid, void *addr )
+{
+    void *base = baseAddress;
+    int i;
+    for(i = 0; iid[i].Characteristics != 0 && iid[i].FirstThunk != 0; i++) {
+        PIMAGE_THUNK_DATA thunkILT = (PIMAGE_THUNK_DATA)(iid[i].Characteristics + base);
+        PIMAGE_THUNK_DATA thunkIAT = (PIMAGE_THUNK_DATA)(iid[i].FirstThunk + base);
+        for(; thunkILT->u1.AddressOfData != 0; thunkILT++, thunkIAT++) {
+            if (IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal))
+              continue;
+            if ((void *)thunkIAT->u1.Function != addr)
+              continue;
+            PIMAGE_IMPORT_BY_NAME nameData = (PIMAGE_IMPORT_BY_NAME)(thunkILT->u1.AddressOfData + base);
+            return nameData->Name;
+        }
+   }
+   return NULL;
+}
+
+/*
+ * Return adress from Image Allocation Table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static unsigned char *getAddressFromIAT( unsigned char *addr )
+{
+    /* ...inline app code...
+     * 00401002  |. E8 7B0D0000    CALL 00401D82               ; \GetModuleHandleA
+     * ...thunk table...
+     * 00401D82   $-FF25 4C204000 , JMP DWORD PTR DS:[40204C]  ;  KERNEL32.GetModuleHandleA
+     * ...memory address value of pointer...
+     * 40204C > FC 3D 57 7C   ;little endian pointer value
+     */
+    if( addr[0] != 0xff || addr[1] != 0x25 )

```

Comment:
`addr` is pointer parameter passed by user from `dladdr()` call. You should check if pointer is valid prior to dereferencing memory as it can cause crash of whole process. I think it should be possible to use `VirtualQuery()` to check if address is valid and check `State` and `Protect` members from returned `MEMORY_BASIC_INFORMATION` structure if from address is possible to read.

Basically calling `dladdr(NULL)` or with any other value, dlfcn-win32 should not crash whole process.

Reply:
> VirtualQuery() 

good point, thanks

> https://docs.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-memory_basic_information

From the doc it looks that  `AllocationProtect` needs to be checked instead. Using invalid addresses also sets `AllocationBase` to NULL, which can be checked also.

Reply:
I think that we should check also for `PAGE_NOACCESS`. As in this case dereferencing pointer would crash process.